File: fxgl/src/main/java/com/almasb/fxgl/gameplay/GameWorld.java
/*
 * The MIT License (MIT)
 *
 * FXGL - JavaFX Game Library
 *
 * Copyright (c) 2015-2017 AlmasB (almaslvl@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.almasb.fxgl.gameplay;

import com.almasb.fxgl.app.FXGL;
import com.almasb.fxgl.core.collection.Array;
import com.almasb.fxgl.core.collection.ObjectMap;
import com.almasb.fxgl.core.reflect.ReflectionUtils;
import com.almasb.fxgl.ecs.Entity;
import com.almasb.fxgl.ecs.EntityWorld;
import com.almasb.fxgl.entity.*;
import com.almasb.fxgl.entity.component.*;
import com.almasb.fxgl.event.EventTrigger;
import com.almasb.fxgl.logging.Logger;
import com.almasb.fxgl.time.UpdateEvent;
import com.almasb.fxgl.time.UpdateEventListener;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.geometry.Point2D;
import javafx.geometry.Rectangle2D;

import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;

/**
 * Represents pure logical state of game.
 * Manages all entities and their state.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
@Singleton
public final class GameWorld extends EntityWorld implements UpdateEventListener {

    private static final Logger log = FXGL.getLogger("FXGL.GameWorld");

    private ObjectProperty<GameDifficulty> gameDifficulty = new SimpleObjectProperty<>(GameDifficulty.MEDIUM);

    /**
     * @return game difficulty
     */
    public GameDifficulty getGameDifficulty() {
        return gameDifficultyProperty().get();
    }

    /**
     * @return game difficulty property
     */
    public ObjectProperty<GameDifficulty> gameDifficultyProperty() {
        return gameDifficulty;
    }

    private Array<EventTrigger<?> > eventTriggers = new Array<>(false, 32);

    @Inject
    protected GameWorld() {
        log.debug("Game world initialized");
    }

    @Override
    public void onUpdateEvent(UpdateEvent event) {
        update(event.tpf());
        updateTriggers(event);
    }

    private void updateTriggers(UpdateEvent event) {
        for (Iterator<EventTrigger<?> > it = eventTriggers.iterator(); it.hasNext(); ) {
            EventTrigger trigger = it.next();

            trigger.onUpdateEvent(event);

            if (trigger.reachedLimit()) {
                it.remove();
            }
        }
    }

    @Override
    public void reset() {
        log.debug("Resetting game world");
        super.reset();
    }

    /**
     * Add event trigger to the world.
     *
     * @param trigger the event trigger
     */
    public void addEventTrigger(EventTrigger<?> trigger) {
        eventTriggers.add(trigger);
    }

    /**
     * Remove event trigger from the world.
     *
     * @param trigger the event trigger
     */
    public void removeEventTrigger(EventTrigger<?> trigger) {
        eventTriggers.removeValue(trigger, true);
    }

    private ObjectProperty<Entity> selectedEntity = new SimpleObjectProperty<>();

    /**
     * Returns last selected (clicked on by mouse) entity.
     *
     * @return selected entity
     */
    public Optional<Entity> getSelectedEntity() {
        return Optional.ofNullable(selectedEntity.get());
    }

    /**
     * @return selected entity property
     */
    public ObjectProperty<Entity> selectedEntityProperty() {
        return selectedEntity;
    }

    /**
     * Set level to given.
     * Resets the world.
     * Adds all level entities to the world.
     *
     * @param level the level
     */
    public void setLevel(Level level) {
        reset();

        log.debug("Setting level: " + level);
        level.getEntities().forEach(this::addEntity);
    }

    private EntityFactory entityFactory = null;

    private ObjectMap<String, Function<SpawnData, Entity>> entitySpawners = new ObjectMap<>();

    /**
     * @return entity factory or null if not set
     */
    @SuppressWarnings("unchecked")
    public <T extends EntityFactory> T getEntityFactory() {
        return (T) entityFactory;
    }

    /**
     * Set main entity factory to be used via {@link GameWorld#spawn(String, SpawnData)}.
     *
     * @param entityFactory factory for creating entities
     */
    public void setEntityFactory(EntityFactory entityFactory) {
        this.entityFactory = entityFactory;

        entitySpawners.clear();

        ReflectionUtils.<SpawnData, Entity, Spawns>findMethodsMapFunctions(entityFactory, Spawns.class)
                .forEach((annotation, spawnerFunction) -> entitySpawners.put(annotation.value(), spawnerFunction));
    }

    /**
     * Creates an entity with given name at x, y using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param position spawn location
     * @return spawned entity
     */
    public Entity spawn(String entityName, Point2D position) {
        return spawn(entityName, position.getX(), position.getY());
    }

    /**
     * Creates an entity with given name at x, y using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param x x position
     * @param y y position
     * @return spawned entity
     */
    public Entity spawn(String entityName, double x, double y) {
        return spawn(entityName, new SpawnData(x, y));
    }

    /**
     * Creates an entity with given name and data using specified entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by {@link Spawns}
     * @param data spawn data, such as x, y and any extra info
     * @return spawned entity
     */
    public Entity spawn(String entityName, SpawnData data) {
        if (entityFactory == null)
            throw new IllegalStateException("EntityFactory was not set!");

        Function<SpawnData, Entity> spawner = entitySpawners.get(entityName);
        if (spawner == null)
            throw new IllegalArgumentException("EntityFactory does not have a method annotated @Spawns(" + entityName + ")");

        Entity entity = spawner.apply(data);
        addEntity(entity);
        return entity;
    }

    /* QUERIES */

    private GameWorldQuery query = new GameWorldQuery(entities);

    /**
     * Returns a list of entities which are filtered by
     * given predicate.
     * Warning: object allocation.
     *
     * @param predicate filter
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesFiltered(Predicate<Entity> predicate) {
        return query.getEntitiesFiltered(predicate);
    }

    /**
     * GC-friendly version of {@link #getEntitiesFiltered(Predicate)}.
     *
     * @param result the array to collect entities
     * @param predicate filter
     */
    public void getEntitiesFiltered(Array<Entity> result, Predicate<Entity> predicate) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (predicate.test(e)) {
                result.add(e);
            }
        }
    }

    /**
     * This query only works on entities with TypeComponent.
     * If called with no arguments, all entities are returned.
     * Warning: object allocation.
     *
     * @param types entity types
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesByType(Enum<?>... types) {
        return query.getEntitiesByType(types);
    }

    /**
     * GC-friendly version of {@link #getEntitiesByType(Enum[])}.
     *
     * @param result the array to collect entities
     * @param types entity types
     */
    public void getEntitiesByType(Array<Entity> result, Enum<?>... types) {
        if (types.length == 0) {
            for (int i = 0; i < entities.size(); i++) {
                Entity e = entities.get(i);
                result.add(e);
            }
            return;
        }

        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            if (isOfType(e, types)) {
                result.add(e);
            }
        }
    }

    private boolean isOfType(Entity e, Enum<?>... types) {
        TypeComponent entityType = Entities.getType(e);
        if (entityType != null) {
            for (Enum<?> type : types) {
                if (entityType.isType(type)) {
                    return true;
                }
            }
        }

        return false;
    }

    /**
     * Returns a list of entities
     * which are partially or entirely
     * in the specified rectangular selection.
     * This query only works on entities with BoundingBoxComponent.
     * Warning: object allocation.
     *
     * @param selection Rectangle2D that describes the selection box
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesInRange(Rectangle2D selection) {
        return query.getEntitiesInRange(selection);
    }

    /**
     * GC-friendly version of {@link #getEntitiesInRange(Rectangle2D)}.
     *
     * @param result the array to collect entities
     * @param minX min x
     * @param minY min y
     * @param maxX max x
     * @param maxY max y
     */
    public void getEntitiesInRange(Array<Entity> result, double minX, double minY, double maxX, double maxY) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            BoundingBoxComponent bbox = Entities.getBBox(e);
            if (bbox != null && bbox.isWithin(minX, minY, maxX, maxY)) {
                result.add(e);
            }
        }
    }

    /**
     * Returns a list of entities
     * which colliding with given entity.
     *
     * Note: CollidableComponent is not considered.
     * This query only works on entities with BoundingBoxComponent.
     *
     * @param entity the entity
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getCollidingEntities(Entity entity) {
        return query.getCollidingEntities(entity);
    }

    /**
     * GC-friendly version of {@link #getCollidingEntities(Entity)}.
     *
     * @param result the array to collect entities
     * @param entity the entity
     */
    public void getCollidingEntities(Array<Entity> result, Entity entity) {
        BoundingBoxComponent entityBBox = Entities.getBBox(entity);

        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);
            BoundingBoxComponent bbox = Entities.getBBox(e);
            if (bbox != null && bbox.isCollidingWith(entityBBox) && e != entity) {
                result.add(e);
            }
        }
    }

    /**
     * Returns a list of entities which have the given render layer index.
     * This query only works on entities with ViewComponent.
     *
     * @param layer render layer
     * @return new list containing entities that satisfy query filters
     */
    public List<Entity> getEntitiesByLayer(RenderLayer layer) {
        return query.getEntitiesByLayer(layer);
    }

    /**
     * GC-friendly version of {@link #getEntitiesByLayer(RenderLayer)}.
     *
     * @param result the array to collect entities
     * @param layer render layer
     */
    public void getEntitiesByLayer(Array<Entity> result, RenderLayer layer) {
        for (int i = 0; i < entities.size(); i++) {
            Entity e = entities.get(i);

            ViewComponent view = Entities.getView(e);

            if (view != null && view.getRenderLayer().index() == layer.index()) {
                result.add(e);
            }
        }
    }

    /**
     * Returns the closest entity to the given entity with given
     * filter. The given
     * entity itself is never returned.
     * <p>
     * If there no entities satisfying the requirement, {@link Optional#empty()}
     * is returned.
     * Warning: object allocation.
     *
     * @param entity selected entity
     * @param filter requirements
     * @return closest entity to selected entity with type
     */
    public Optional<Entity> getClosestEntity(Entity entity, Predicate<Entity> filter) {
        Array<Entity> array = new Array<>(false, 64);

        for (Entity e : getEntitiesByComponent(PositionComponent.class)) {
            if (filter.test(e) && e != entity) {
                array.add(e);
            }
        }

        if (array.size() == 0)
            return Optional.empty();

        array.sort((e1, e2) -> (int) (Entities.getPosition(e1).distance(Entities.getPosition(entity))
                        - Entities.getPosition(e2).distance(Entities.getPosition(entity))));

        return Optional.of(array.get(0));
    }

    /**
     * Returns an entity at given position. The position x and y
     * must equal to entity's position x and y.
     * <p>
     * Returns {@link Optional#empty()} if no entity was found at
     * given position.
     * This query only works on entities with PositionComponent.
     *
     * @param position point in the world
     * @return entity at point
     */
    public Optional<Entity> getEntityAt(Point2D position) {
        for (Entity e : getEntitiesByComponent(PositionComponent.class)) {
            if (Entities.getPosition(e).getValue().equals(position)) {
                return Optional.of(e);
            }
        }

        return Optional.empty();
    }

    /**
     * Returns an entity whose IDComponent matches given name and id.
     * <p>
     * Returns {@link Optional#empty()} if no entity was found with such combination.
     * This query only works on entities with IDComponent.
     *
     * @param name entity name
     * @param id entity id
     * @return entity that matches the query or {@link Optional#empty()}
     */
    public Optional<Entity> getEntityByID(String name, int id) {
        for (Entity e : getEntitiesByComponent(IDComponent.class)) {
            IDComponent idComponent = e.getComponentUnsafe(IDComponent.class);
            if (idComponent.getName().equals(name) && idComponent.getID() == id) {
                return Optional.of(e);
            }
        }

        return Optional.empty();
    }
}

File: fxgl-base/src/main/kotlin/com/almasb/fxgl/entity/GameWorld.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity

import com.almasb.fxgl.app.FXGL
import com.almasb.fxgl.app.tryCatchRoot
import com.almasb.fxgl.core.collection.Array
import com.almasb.fxgl.core.collection.ObjectMap
import com.almasb.fxgl.core.collection.UnorderedArray
import com.almasb.fxgl.core.logging.Logger
import com.almasb.fxgl.core.math.FXGLMath
import com.almasb.fxgl.core.reflect.ReflectionUtils
import com.almasb.fxgl.entity.component.Component
import com.almasb.fxgl.entity.components.IDComponent
import com.almasb.fxgl.entity.components.IrremovableComponent
import com.almasb.fxgl.entity.components.TimeComponent
import com.almasb.fxgl.parser.tiled.TiledMap
import com.almasb.fxgl.util.Optional
import com.almasb.fxgl.util.Predicate
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.geometry.Point2D
import javafx.geometry.Rectangle2D

/**
 * Represents pure logical state of the game.
 * Manages all entities and allows queries.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
class GameWorld {

    companion object {
        private const val INITIAL_CAPACITY = 32

        private val log = Logger.get("GameWorld")
    }

    private val updateList = Array<Entity>(INITIAL_CAPACITY)

    /**
     * List of entities added to the update list in the next tick.
     */
    private val waitingList = UnorderedArray<Entity>(INITIAL_CAPACITY)

    /**
     * List of entities in the world.
     *
     * @return direct list of entities in the world (do NOT modify)
     */
    val entities = ArrayList<Entity>(INITIAL_CAPACITY)

    /**
     * @return shallow copy of the entities list (new list)
     */
    val entitiesCopy: List<Entity>
        get(): List<Entity> = ArrayList(entities)

    init {
        log.debug("Game world initialized")
    }

    /**
     * The entity will be added to update list in the next tick.
     *
     * @param entity the entity to add to world
     */
    fun addEntity(entity: Entity) {
        if (entity.isActive)
            throw IllegalArgumentException("Entity is already attached to world")

        waitingList.add(entity)
        entities.add(entity)

        add(entity)
    }

    fun addEntities(vararg entitiesToAdd: Entity) {
        for (e in entitiesToAdd) {
            addEntity(e)
        }
    }

    private fun add(entity: Entity) {
        entity.init(this)
        notifyEntityAdded(entity)
    }

    /**
     * Entity will be removed from world and parties notified of removal in the same frame.
     * The components and controls of entity will be removed in the next frame to avoid
     * concurrency issues.
     */
    fun removeEntity(entity: Entity) {
        if (!entity.isActive) {
            log.warning("Attempted to remove entity which is not active")
            return
        }

        if (!canRemove(entity))
            return

        if (entity.world !== this)
            throw IllegalArgumentException("Attempted to remove entity not attached to this world")

        entities.remove(entity)

        entity.markForRemoval()
        notifyEntityRemoved(entity)

        // we cannot clean entities here because this may have been called through a control
        // while entity is being updated
        // so, we clean the entity on next frame
    }

    fun removeEntities(vararg entitiesToRemove: Entity) {
        for (e in entitiesToRemove) {
            removeEntity(e)
        }
    }

    fun removeEntities(entitiesToRemove: Collection<Entity>) {
        for (e in entitiesToRemove) {
            removeEntity(e)
        }
    }

    private fun canRemove(entity: Entity): Boolean {
        return !entity.hasComponent(IrremovableComponent::class.java)
    }

    /**
     * Performs a single world update tick.
     *
     * @param tpf time per frame
     */
    fun onUpdate(tpf: Double) {
        updateList.addAll(waitingList)
        waitingList.clear()

        val it = updateList.iterator()
        while (it.hasNext()) {
            val e = it.next()

            if (!e.isActive) {
                // clean entities removed in the last frame
                e.clean()
                it.remove()
            } else {
                val tpfRatio = e.getComponentOptional(TimeComponent::class.java)
                        .map { it.value }
                        .orElse(1.0)

                e.update(tpf * tpfRatio)
            }
        }
    }

    /**
     * Removes all (including with IrremovableComponent) entities.
     * Does NOT clear state listeners.
     * Do NOT call this method manually.
     * It is called automatically by FXGL during initGame().
     */
    fun clear() {
        log.debug("Clearing game world")

        waitingList.clear()

        // we may still have some entities that have been removed but not yet cleaned
        run {
            val it = updateList.iterator()
            while (it.hasNext()) {
                val e = it.next()

                if (!e.isActive) {
                    e.clean()
                }

                it.remove()
            }
        }

        // entities list does not contain "not active" entities, so we do full clean
        val it = entities.iterator()
        while (it.hasNext()) {
            val e = it.next()

            e.markForRemoval()
            notifyEntityRemoved(e)
            e.clean()

            it.remove()
        }

        entityFactories.clear()
        entitySpawners.clear()
    }

    private val worldListeners = Array<EntityWorldListener>()

    fun addWorldListener(listener: EntityWorldListener) {
        worldListeners.add(listener)
    }

    fun removeWorldListener(listener: EntityWorldListener) {
        worldListeners.removeValueByIdentity(listener)
    }

    private fun notifyEntityAdded(e: Entity) {
        for (i in 0 until worldListeners.size()) {
            worldListeners.get(i).onEntityAdded(e)
        }
    }

    private fun notifyEntityRemoved(e: Entity) {
        for (i in 0 until worldListeners.size()) {
            worldListeners.get(i).onEntityRemoved(e)
        }
    }

    private val selectedEntity = SimpleObjectProperty<Entity>()

    /**
     * @return last selected (clicked on by mouse) entity
     */
    fun getSelectedEntity(): Optional<Entity> {
        return Optional.ofNullable(selectedEntity.get())
    }

    /**
     * @return selected entity property
     */
    fun selectedEntityProperty(): ObjectProperty<Entity> {
        return selectedEntity
    }

    /**
     * Set level to given.
     * Resets the world.
     * Adds all level entities to the world.
     *
     * @param level the level
     */
    fun setLevel(level: Level) {
        clearLevel()

        log.debug("Setting level: $level")
        level.entities.forEach { addEntity(it) }
    }

    /**
     * @param mapFileName name of the .json file or the .tmx file
     */
    fun setLevelFromMap(mapFileName: String) {
        if (mapFileName.endsWith(".json")) {
            setLevelFromMap(FXGL.getAssetLoader().loadJSON(mapFileName, TiledMap::class.java))
        } else if (mapFileName.endsWith(".tmx")) {
            setLevelFromMap(FXGL.getAssetLoader().loadTMX(mapFileName))
        } else {
            throw IllegalArgumentException("Unknown Tiled map format")
        }
    }

    fun setLevelFromMap(map: TiledMap) {
        clearLevel()

        log.debug("Setting level from map")

        map.layers.filter { it.type == "tilelayer" }
                .forEach {
                    Entities.builder()
                            .viewFromTiles(map, it.name, RenderLayer.BACKGROUND)
                            .buildAndAttach(this)
                }

        map.layers.filter { it.type == "objectgroup" }
                .flatMap { it.objects }
                .forEach { obj ->
                    val data = SpawnData(obj)

                    data.put("tilesets", map.tilesets)

                    spawn(obj.type, data)
                }
    }

    /**
     * Removes removable entities.
     */
    private fun clearLevel() {
        log.debug("Clearing removable entities")

        waitingList.clear()

        // we may still have some entities that have been removed but not yet cleaned
        // but we do not want to remove Irremovables
        run {
            val it = updateList.iterator()
            while (it.hasNext()) {
                val e = it.next()

                if (canRemove(e)) {
                    if (!e.isActive) {
                        // clean it here because "entities" list does not have "e"
                        e.clean()
                    }

                    it.remove()
                }
            }
        }

        // entities list does not contain "not active" entities, so we do full clean
        // but we do not want to remove Irremovables
        val it = entities.iterator()
        while (it.hasNext()) {
            val e = it.next()

            if (canRemove(e)) {
                e.markForRemoval()
                notifyEntityRemoved(e)
                e.clean()

                it.remove()
            }
        }
    }

    private val entityFactories = ObjectMap<EntityFactory, List<String>>()
    private val entitySpawners = ObjectMap<String, EntitySpawner>()

    /**
     * @param entityFactory factory for creating entities
     */
    fun addEntityFactory(entityFactory: EntityFactory) {
        val entityNames = arrayListOf<String>()

        ReflectionUtils.findMethodsMapToFunctions(entityFactory, Spawns::class.java, EntitySpawner::class.java)
                .forEach { annotation, entitySpawner ->

                    val entityName = annotation.value

                    checkDuplicateSpawners(entityFactory, entityName)

                    entitySpawners.put(entityName, entitySpawner)
                    entityNames.add(entityName)
                }

        entityFactories.put(entityFactory, entityNames)
    }

    private fun checkDuplicateSpawners(entityFactory: EntityFactory, entityName: String) {
        if (entitySpawners.containsKey(entityName)) {

            // find the factory that already has entityName spawner
            val factory = entityFactories.find { it.value.contains(entityName) }

            throw IllegalArgumentException("Duplicated @Spawns($entityName) in $entityFactory. Already exists in $factory")
        }
    }

    fun removeEntityFactory(entityFactory: EntityFactory) {
        entityFactories.remove(entityFactory)?.forEach {
            entitySpawners.remove(it)
        }
    }

    /**
     * Creates an entity with given name at 0, 0 using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @return spawned entity
     */
    fun spawn(entityName: String): Entity {
        return spawn(entityName, 0.0, 0.0)
    }

    /**
     * Creates an entity with given name at x, y using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param position spawn location
     * @return spawned entity
     */
    fun spawn(entityName: String, position: Point2D): Entity {
        return spawn(entityName, position.x, position.y)
    }

    /**
     * Creates an entity with given name at x, y using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param x x position
     * @param y y position
     * @return spawned entity
     */
    fun spawn(entityName: String, x: Double, y: Double): Entity {
        return spawn(entityName, SpawnData(x, y))
    }

    /**
     * Creates an entity with given name and data using a previously added entity factory.
     * Adds created entity to this game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param data spawn data, such as x, y and any extra info
     * @return spawned entity
     */
    fun spawn(entityName: String, data: SpawnData): Entity {
        val entity = create(entityName, data)
        addEntity(entity)
        return entity
    }

    /**
     * Creates an entity with given name and data using a previously added entity factory.
     * Does NOT add created entity to the game world.
     *
     * @param entityName name of entity as specified by [Spawns]
     * @param data spawn data, such as x, y and any extra info
     * @return created entity
     */
    fun create(entityName: String, data: SpawnData): Entity {
        if (entityFactories.size() == 0)
            throw IllegalStateException("No EntityFactory was added! Call gameWorld.addEntityFactory()")

        val spawner = entitySpawners.get(entityName)
                ?: throw IllegalArgumentException("No EntityFactory has a method annotated @Spawns($entityName)")

        if (!data.hasKey("type")) {
            data.put("type", entityName)
        }

        return tryCatchRoot { spawner.apply(data) }

//        try {
//            return spawner.apply(data)
//        } catch (e: Exception) {
//            throw ReflectionUtils.getRootCause(e)
//        }
    }

    /* QUERIES */

    /**
     * @return a group of entities of particular type, this is preferred over querying
     * manually if called many times
     */
    fun <T : Entity> getGroup(vararg types: Enum<*>): EntityGroup<T> {
        return EntityGroup(this, getEntitiesByType(*types) as List<T>, *types)
    }

    /**
     * Useful for singleton type entities, e.g. Player.
     *
     * @return first occurrence matching given type
     */
    fun getSingleton(type: Enum<*>): Optional<Entity> {
        return getSingleton(Predicate { it.isType(type) })
    }

    /**
     * @return first occurrence matching given predicate
     */
    fun getSingleton(predicate: Predicate<Entity>): Optional<Entity> {
        return Optional.ofNullable(entities.find { predicate.test(it) })
    }

    /**
     * @param type entity type
     * @return a random entity with given type
     */
    fun getRandom(type: Enum<*>): Optional<Entity> {
        return FXGLMath.random(getEntitiesByType(type))
    }

    fun getRandom(predicate: Predicate<Entity>): Optional<Entity> {
        return FXGLMath.random(getEntitiesFiltered(predicate))
    }

    /**
     * @param type component type
     * @return array of entities that have given component
     */
    fun getEntitiesByComponent(type: Class<out Component>): List<Entity> {
        return entities.filter { it.hasComponent(type) }
    }

    /**
     * Returns a list of entities which are filtered by
     * given predicate.
     * Warning: object allocation.
     *
     * @param predicate filter
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesFiltered(predicate: Predicate<Entity>): List<Entity> {
        return entities.filter { predicate.test(it) }
    }

    /**
     * GC-friendly version of [.getEntitiesFiltered].
     *
     * @param result the array to collect entities
     * @param predicate filter
     */
    fun getEntitiesFiltered(result: Array<Entity>, predicate: Predicate<Entity>) {
        for (i in entities.indices) {
            val e = entities[i]
            if (predicate.test(e)) {
                result.add(e)
            }
        }
    }

    /**
     * If called with no arguments, all entities are returned.
     * Warning: object allocation.
     *
     * @param types entity types
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesByType(vararg types: Enum<*>): List<Entity> {
        if (types.isEmpty())
            return entitiesCopy

        return entities.filter { isOneOfTypes(it, *types) }
    }

    private fun isOneOfTypes(entity: Entity, vararg types: Enum<*>): Boolean {
        return types.any { entity.isType(it) }
    }

    /**
     * GC-friendly version of [.getEntitiesByType].
     *
     * @param result the array to collect entities
     * @param types entity types
     */
    fun getEntitiesByType(result: Array<Entity>, vararg types: Enum<*>) {
        if (types.size == 0) {
            for (i in entities.indices) {
                val e = entities[i]
                result.add(e)
            }
            return
        }

        for (i in entities.indices) {
            val e = entities[i]
            if (isOfType(e, *types)) {
                result.add(e)
            }
        }
    }

    private fun isOfType(e: Entity, vararg types: Enum<*>): Boolean {
        for (type in types) {
            if (e.isType(type)) {
                return true
            }
        }

        return false
    }

    /**
     * Returns a list of entities
     * which are partially or entirely
     * in the specified rectangular selection.
     * Warning: object allocation.
     *
     * @param selection Rectangle2D that describes the selection box
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesInRange(selection: Rectangle2D): List<Entity> {
        return entities.filter { it.boundingBoxComponent.isWithin(selection) }
    }

    /**
     * GC-friendly version of [.getEntitiesInRange].
     *
     * @param result the array to collect entities
     * @param minX min x
     * @param minY min y
     * @param maxX max x
     * @param maxY max y
     */
    fun getEntitiesInRange(result: Array<Entity>, minX: Double, minY: Double, maxX: Double, maxY: Double) {
        for (i in entities.indices) {
            val e = entities[i]
            if (e.boundingBoxComponent.isWithin(minX, minY, maxX, maxY)) {
                result.add(e)
            }
        }
    }

    /**
     * Returns a list of entities
     * which colliding with given entity.
     *
     * Note: CollidableComponent is not considered.
     *
     * @param entity the entity
     * @return new list containing entities that satisfy query filters
     */
    fun getCollidingEntities(entity: Entity): List<Entity> {
        return entities.filter { it.isColliding(entity) && it !== entity }
    }

    /**
     * GC-friendly version of [.getCollidingEntities].
     *
     * @param result the array to collect entities
     * @param entity the entity
     */
    fun getCollidingEntities(result: Array<Entity>, entity: Entity) {
        val entityBBox = entity.boundingBoxComponent

        for (i in entities.indices) {
            val e = entities[i]
            if (e.boundingBoxComponent.isCollidingWith(entityBBox) && e !== entity) {
                result.add(e)
            }
        }
    }

    /**
     * Returns a list of entities which have the given render layer index.
     *
     * @param layer render layer
     * @return new list containing entities that satisfy query filters
     */
    fun getEntitiesByLayer(layer: RenderLayer): List<Entity> {
        return entities.filter { it.renderLayer.index() == layer.index() }
    }

    /**
     * GC-friendly version of [.getEntitiesByLayer].
     *
     * @param result the array to collect entities
     * @param layer render layer
     */
    fun getEntitiesByLayer(result: Array<Entity>, layer: RenderLayer) {
        for (i in entities.indices) {
            val e = entities[i]

            if (e.renderLayer.index() == layer.index()) {
                result.add(e)
            }
        }
    }

    /**
     * Returns a list of entities at given position.
     * The position x and y must equal to entity's position x and y.
     *
     * @param position point in the world
     * @return entities at given point
     */
    fun getEntitiesAt(position: Point2D): List<Entity> {
        return entities.filter { it.position == position }
    }

    /**
     * GC-friendly version of [.getEntitiesAt].
     *
     * @param result the array to collect entities
     * @param position point in the world
     */
    fun getEntitiesAt(result: Array<Entity>, position: Point2D) {
        for (i in entities.indices) {
            val e = entities[i]

            if (e.position == position) {
                result.add(e)
            }
        }
    }

    /**
     * Returns the closest entity to the given entity with given
     * filter. The given
     * entity itself is never returned.
     *
     *
     * If there no entities satisfying the requirement, [Optional.empty]
     * is returned.
     * Warning: object allocation.
     *
     * @param entity selected entity
     * @param filter requirements
     * @return closest entity to selected entity with type
     */
    fun getClosestEntity(entity: Entity, filter: Predicate<Entity>): Optional<Entity> {
        val array = UnorderedArray<Entity>(64)

        for (e in entities) {
            if (filter.test(e) && e !== entity) {
                array.add(e)
            }
        }

        if (array.size() == 0)
            return Optional.empty()

        array.sort { e1, e2 -> (e1.distance(entity) - e2.distance(entity)).toInt() }

        return Optional.of(array.get(0))
    }

    /**
     * Returns an entity whose IDComponent matches given name and id.
     *
     *
     * Returns [Optional.empty] if no entity was found with such combination.
     * This query only works on entities with IDComponent.
     *
     * @param name entity name
     * @param id entity id
     * @return entity that matches the query or [Optional.empty]
     */
    fun getEntityByID(name: String, id: Int): Optional<Entity> {
        for (e in getEntitiesByComponent(IDComponent::class.java)) {
            val idComponent = e.getComponent(IDComponent::class.java)
            if (idComponent.name == name && idComponent.id == id) {
                return Optional.of(e)
            }
        }

        return Optional.empty()
    }
}